opencv/c_extension/fast_image_processing.cpp

```diff
@@ -0,0 +1,229 @@
+#include <Eigen/Dense>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/core/eigen.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+
+#define CONVERT_TO_INT true
+#define MAX_WIDTH 320
+#define PI 3.14159265
+#define REF_ANGLE float(- PI / 2)
+#define MAX_ANGLE float(PI / 4)
+#define IM_WIDTH MAX_WIDTH
+#define N_REGIONS 3
+
+#include "ndarray_converter.h"
+
+using namespace Eigen;
+using EigenMat = Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+
+namespace py = pybind11;
+
+bool initialized = false;
+
+
+cv::Mat centroids;
+cv::Mat resized = cv::Mat::ones(20, 80, CV_32FC3);
+
+cv::Mat w1 = cv::Mat::ones(80*20*3, 20, CV_32F).t();
+cv::Mat b1 = cv::Mat::ones(20, 1, CV_32F).t();
+
+cv::Mat w2 = cv::Mat::ones(20, 4, CV_32F).t();
+cv::Mat b2 = cv::Mat::ones(4, 1, CV_32F).t();
+
+cv::Mat w3 = cv::Mat::ones(4, 1, CV_32F).t();
+cv::Mat b3 = cv::Mat::ones(1, 1, CV_32F).t();
+
+
+// Regions of interest (ROIs)
+cv::Rect R0(0, 150, MAX_WIDTH, 50);
+cv::Rect R1(0, 125, MAX_WIDTH, 50);
+cv::Rect R2(0, 100, MAX_WIDTH, 50);
+cv::Rect R3(0, 75, MAX_WIDTH, 50);
+cv::Rect R4(0, 50, MAX_WIDTH, 50);
+// ROI used
+std::vector<cv::Rect> REGIONS({R1, R2, R3});
+
+/**
+ * Set the weights of the neural network
+ * it should be called from python
+ */
+void setWeights(cv::Mat w1_, cv::Mat b1_, cv::Mat w2_, cv::Mat b2_, cv::Mat w3_, cv::Mat b3_)
+{
+  w1 = w1_.clone().t();
+  w2 = w2_.clone().t();
+  w3 = w3_.clone().t();
+  b1 = b1_.clone().t();
+  b2 = b2_.clone().t();
+  b3 = b3_.clone().t();
+  // Concatenate bias to avoid broadcast issue
+  // (addition between a matrix and a vector)
+  // TODO: replace vector creation with
+  //  a loop that depends on the number of regions
+  CV_Assert(REGIONS.size() == N_REGIONS);
+  std::vector<cv::Mat> m1 {b1, b1, b1};
+  cv::vconcat(m1, b1);
+  std::vector<cv::Mat> m2 {b2, b2, b2};
+  cv::vconcat(m2, b2);
+  std::vector<cv::Mat> m3 {b3, b3, b3};
+  cv::vconcat(m3, b3);
+  initialized = true;
+}
+
+/**
+ * Reshape and normalize an input image
+ * @param  image CV_32FC3
+ */
+cv::Mat preprocessImage(cv::Mat image)
+{
+  // WARNING: resizing is a bottleneck in the computation
+  // Best image quality: cv::INTER_AREA
+  // Fastest:  cv::INTER_NEAREST
+  // Between the two: cv::INTER_LINEAR (bilinear interpolation)
+  cv::resize(image, image, resized.size(), 0, 0, cv::INTER_NEAREST);
+  image = image.reshape(1, 1);
+  image /= 255.0;
+  image -= 0.5;
+  image *= 2;
+  return image;
+}
+
+/**
+ * Rectifier activation
+ */
+float relu(float x)
+{
+    return std::max(float(0.0), x);
+}
+
+/**
+ * Forward pass of the neural network
+ * @param  x CV_32F
+ * @return   CV_32F
+ */
+cv::Mat forward(cv::Mat x)
+{
+  CV_Assert(x.type() == CV_32F);
+
+  // Convert to eigen matrices
+  Map<EigenMat> w1_eigen(w1.ptr<float>(), w1.rows, w1.cols);
+  Map<EigenMat> b1_eigen(b1.ptr<float>(), b1.rows, b1.cols);
+  Map<EigenMat> w2_eigen(w2.ptr<float>(), w2.rows, w2.cols);
```

Comment:
Python style Dictionaries everywhere! is really needed?

Reply:
It avoids copy of memory and does not give any seg fault, i did not find a better way yet